### ISSUE TITLE
Legends for scatter plot cannot be created when not using numbers for argument "c"

### ISSUE THREAD

**Issue by siheming** [matplotlib/matplotlib#16616]
### Bug report

**Bug summary**

When trying to create a scatter plot with custom colors the examples for creating a legend do not work anymore.  
If I missed a relevant example please let me know

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

N = 45
x, y = np.random.rand(2, N)
c = np.random.randint(1, 5, size=N)
color  = [f"C{i}" for i in c]
s = np.random.randint(10, 220, size=N)

fig, ax = plt.subplots()

scatter1 = ax.scatter(x, y, c=color)
scatter2 = ax.scatter(y,x, c=c)

legend1 = ax.legend(*scatter1.legend_elements(), loc="upper left", title="Classes1")
legend2 = ax.legend(*scatter2.legend_elements(), loc="lower right", title="Classes2")

ax.add_artist(legend1)
ax.add_artist(legend2)
```
  
**Output**
The first legend does not contain anything, since the legend_elements of scatter1 is empty.

**Matplotlib version**

  * Operating system: Windows 10
  * Matplotlib version: 3.1.1
  * Matplotlib backend (`print(matplotlib.get_backend())`): 'module://ipykernel.pylab.backend_inline'
  * Python version: 3.7.5
  * Jupyter version (if applicable): 
jupyter --version
jupyter core     : 4.5.0
jupyter-notebook : 6.0.1
qtconsole        : 4.5.5
ipython          : 7.8.0
ipykernel        : 5.1.2
jupyter client   : 5.3.3
jupyter lab      : 1.1.4
nbconvert        : 5.6.0
ipywidgets       : 7.5.1
nbformat         : 4.4.0
traitlets        : 4.3.3
  * Other libraries: 

used conda default channel

**Comment by ImportanceOfBeingErnest** [comment]
I guess you meant to say that *the legend_elements of scatter1 is empty*, right? That is expected, and there is a warning about it. 

> \matplotlib\collections.py:960: UserWarning: Collection without array used. Make sure to specify the values to be colormapped via the `c` argument.

The reason is that in case you specify the scatter colors as a simple list of colors, no mapping is happening, and therefore we cannot know what to show in the legend (other than maybe the color itself?).  
Possibly we might think about extending the functionality, but for this it would be good to collect use cases. I.e. can you tell what exactly you would expect the legend to show in the above example case?

**Comment by siheming** [comment]
> I guess you meant to say that _the legend_elements of scatter1 is empty_, right? 

yes, i updated it. my example was slightly different before posting.
> That is expected, and there is a warning about it.
> > \matplotlib\collections.py:960: UserWarning: Collection without array used. Make sure to specify the values to be colormapped via the `c` argument.

In what other manner could I use the `c` argument then? I found this warning not helpful because I was using the same parameter in both examples
> 
> The reason is that in case you specify the scatter colors as a simple list of colors, no mapping is happening, and therefore we cannot know what to show in the legend (other than maybe the color itself?).
> Possibly we might think about extending the functionality, but for this it would be good to collect use cases. I.e. can you tell what exactly you would expect the legend to show in the above example case?

What would be helpful for me is that the list of lines is returned so i can map the labels by myself.
  
   
But maybe I am just using scatter incorretly in this instace. What I would like to do is create a scatter plot with custom colors for every class, with a corretly labeled legend and not use loops and the `label` argument. Is it possible to provide a list of classes (instead of colors) and then pass a dict like mapping with class -> color?

**Comment by ImportanceOfBeingErnest** [comment]
It looks like you want to use a `BoundaryNorm`. An easy option to get that is via [`matplotlib.colors.from_levels_and_colors`](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.colors.from_levels_and_colors.html), like this:

```
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors

N = 45
x, y = np.random.rand(2, N)

c = np.random.randint(1, 5, size=N)
colors  = [f"C{i}" for i in np.arange(1, c.max()+1)]
cmap, norm = matplotlib.colors.from_levels_and_colors(np.arange(1, c.max()+2), colors)

fig, ax = plt.subplots()

scatter1 = ax.scatter(x, y, c=c, cmap=cmap, norm=norm)
ax.legend(*scatter1.legend_elements(), loc="upper left", title="Classes1")

plt.show()
```

![image](https://user-images.githubusercontent.com/23121882/75630970-17090580-5bef-11ea-9697-8f6202fb5f29.png)

There is currently no example showing that, especially not in conjunction with `scatter`. Maybe it's worthwhile to add such example.

Concerning the warning, possibly it can be made more clear; proposals are welcome.

**Comment by siheming** [comment]
Thanks for the example.
I was trying to have labels that are not just numbers but actual classes. I adjusted from the example to the direction I am coming from with my data, however it feels rather cluncy to me (and it also looks like it could easily break if the order of the labels is changed in between). Is there another convenience function that I am missing?

```python
from sklearn.datasets import load_iris
import pandas as pd
import matplotlib.pyplot as plt

iris = load_iris()
df = pd.DataFrame(iris['data'], columns= iris['feature_names'],
                  index=[iris["target_names"][t] for t in iris["target"]])

colors_map = {idx: i for i, idx in enumerate(df.index.unique())}
colors_map_reverse = {i: idx for i, idx in enumerate(df.index.unique())}

c = [colors_map[idx] for idx in df.index]


colors = [f"C{i}" for i, idx in enumerate(df.index.unique())]
cmap, norm = matplotlib.colors.from_levels_and_colors(range(min(c), max(c)+2), colors)

fig, ax = plt.subplots()
scatter1 = ax.scatter(df["sepal length (cm)"], df["sepal width (cm)"], c=c, cmap=cmap, norm=norm)
ax.legend(scatter1.legend_elements()[0], [colors_map_reverse[i] for i in range(3)],
          loc="upper left", title="Classes")
```

> Concerning the warning, possibly it can be made more clear; proposals are welcome.

I will think about what exactly threw me off.

**Comment by siheming** [comment]
There are two different ways that came to mind about how I would like to use scatter to create the same figure as before.

```python
# code from before
df = ...
label_colors = {idx: f"C{i}" for i, idx in enumerate(df.index.unique())}
# could also not be kwargs, but I am not sure what this function would return
some_kwargs_for_scatter = plt.some_helper_for_scatter(df.index, label_colors)

fig, ax = plt.subplots()
scatter1 = ax.scatter(df["sepal length (cm)"], df["sepal width (cm)"], **some_kwargs_for_scatter)
ax.legend(*scatter1.legend_elements(),
          loc="upper left", title="Classes")
```

or just passing it to scatter

```python
# code from before
df = ...
label_colors = {idx: f"C{i}" for i, idx in enumerate(df.index.unique())}

fig, ax = plt.subplots()
scatter1 = ax.scatter(df["sepal length (cm)"], df["sepal width (cm)"], c_labels=df.index, label_colors=label_colors)
ax.legend(*scatter1.legend_elements(),
          loc="upper left", title="Classes")
```

Is there any straightforward way to achieve this at the moment?

**Comment by siheming** [comment]
@ImportanceOfBeingErnest did you get a chance to look at my updated example, especially the labels being the strings of the classes instead of the numbers that were passed with the `c` argument.

Is there a function which solves this issue for scatter plots?

**Comment by tacaswell** [comment]
You can do

```python
scatter1 = ax.scatter(x, y, c=color, label="label1")
scatter2 = ax.scatter(y,x, c=c, label="label2")
ax.legend()
```

which I think will get what you want.  

The `scatter2.legend_elements()` helper is designed to make the case where you are doing color mapping inside of scatter easier.

**Comment by siheming** [comment]
@tacaswell I'd like to avoid that because for my acutal usecase i have lots of subplots (50-200) and up to 20 differently labeled set of points. I have it currently implemented like you suggest and it is terribly slow. so I want to get away from that

**Comment by tacaswell** [comment]
Interesting, have you profiled what part if it is slow?  My knee-jerk guess is that the main source of the slowness in the many subplots rather than many scatters.

**Comment by siheming** [comment]
No, I have not profiled it yet. My thought was just "Double for-loop, can't expect any decent performance", but if you think that the issue is elsewhere I might test that. If the bottleneck are the subplots then improving performance will not go very far if I understand correctly?

**Comment by jklymak** [comment]
Just do it with your 20 scatters and without and see what the difference is.  My hunch would also be many subplots.  Particularly if you use constrained_layout or tight_layout neither of which are very fast with lots of subplots.

**Comment by ImportanceOfBeingErnest** [comment]
One might be tempted to think that creating several artists in a loop is slower than one single scatter. Interestingly, this is not necessarily the case. Comparing the case where individual scatters per category are created to a single scatter shows that using the loop is faster, as long as there aren't too many categories. 

Leaving out the boilerplate needed for timing, consider
```
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import from_levels_and_colors

def get_data(n, n_cats):
    x, y = np.random.rand(2,n)
    c = np.random.randint(0, n_cats, n).astype(str)
    return x,y,c

def loop_simple(x,y,c, show=False):
    u, inv = np.unique(c, return_inverse=True)
        
    fig, ax = plt.subplots()
    for cat, color in zip(u, plt.cm.viridis(np.linspace(0,1,len(u)))):
        inx = np.where(c==cat)
        ax.scatter(x[inx], y[inx], color=color, label=cat)
    ax.legend()

        
def loop_simple_plot(x,y,c):
    u, inv = np.unique(c, return_inverse=True)
        
    fig, ax = plt.subplots()
    for cat, color in zip(u, plt.cm.viridis(np.linspace(0,1,len(u)))):
        inx = np.where(c==cat)
        ax.plot(x[inx], y[inx], color=color, ls="", marker="o", label=cat)
    ax.legend()


def single_scatter(x,y,c):   
    u, inv = np.unique(c, return_inverse=True)
    cmap, norm = from_levels_and_colors(np.arange(0,len(u)+1)-0.5, plt.cm.viridis(np.linspace(0,1,len(u))))
    
    fig, ax = plt.subplots()
    sc = ax.scatter(x, y, c=inv, cmap=cmap, norm=norm)
    
    ax.legend(sc.legend_elements()[0], u)

```

![image](https://user-images.githubusercontent.com/23121882/76153618-c140cb80-60ce-11ea-90ea-308e0fb081fa.png)

![image](https://user-images.githubusercontent.com/23121882/76153620-c867d980-60ce-11ea-8843-ea09faca4548.png)

![image](https://user-images.githubusercontent.com/23121882/76153622-cef65100-60ce-11ea-9f82-dffeebfacdb1.png)

**Comment by siheming** [comment]
Thanks for the very extensive answers. I suspect that based on the information from the last two answers the speed issue in my actual code probably comes from using the thight layout, which I will test when I get the time to. Thanks alot again!
